### PR TITLE
Add support for Xcode 11 new xcresult format

### LINKED ISSUE #32: Support for Xcode 11
Xcode 11 uses a new format for `xctest` (version 3), which doesn't have a `TestSummary.plist` file, and uses Zstandard-compressed "database-like" files.

Xcode 11 Release Notes state:

> The format of result bundles changed in Xcode 11. A result bundle is an asset produced by Xcode 11 with the `xcresult` file extension that contains information about the build, tests, code coverage, and more. Any `xcresult` files produced with Xcode 10 or earlier cannot be read by Xcode 11. A result bundle can be produced by passing `-resultBundlePath ./Example.xcresult` to an `xcodebuild` invocation and the `Example.xcresult` can then be opened in Xcode. Xcode also creates result bundles in `Derived Data`. The current result bundle version number is 3, which can be specified by passing the `xcodebuild` flag `-resultBundleVersion 3`. Version 3 is the default in Xcode 11, but it is still recommended for automation to explicitly pass the flag, so that any potential future versions that become the default do not cause issues to existing tools. Result bundles can be inspected using `xcresulttool`. A JSON representation of the root object of the result bundle can be exported using `xcrun xcresulttool get --format json --path ./Example.xcresult` and any nested object, identified by its reference found in the JSON output, can be exported by adding the flag `--id REF`. `xcresulttool` also provides the description of its format using `xcrun xcresulttool formatDescription`. (41633595)

Basically, to obtain results from an Xcode 11 `xcresult`, one would have to do the following:

```
// The output of this command will be useful to automatically parse the output of the next commands
$ xcrun xcresulttool formatDescription get --format json
[...]
```

```
$ xcrun xcresulttool get --format json --path Project.xcresult
[...]
          "testsRef" : {
            "_type" : {
              "_name" : "Reference"
            },
            "id" : {
              "_type" : {
                "_name" : "String"
              },
              "_value" : "0~YVEC1htpkY3jLmeu4DB8ah-sJueP7_vUdqA2rL1_R3xT_aftqqEgGUSCHZ0mmpojBSyy4mEL4LX2xyFlfr4pEQ=="
            },
[...]
```

```
// We take the value of the `testsRef` key (if any) for each `ActionRecord` entry
$ xcrun xcresulttool get --format json --path Project.xcresult --id '0~YVEC1htpkY3jLmeu4DB8ah-sJueP7_vUdqA2rL1_R3xT_aftqqEgGUSCHZ0mmpojBSyy4mEL4LX2xyFlfr4pEQ=='
{
  "_type" : {
    "_name" : "ActionTestPlanRunSummaries"
  },
  "summaries" : {
    "_type" : {
      "_name" : "Array"
    },
    "_values" : [
[...]
```

### PR DESCRIPTION
**Pull request by joshdholtz**
Fixes #32

## Motivation
Support Xcode 11 xcresult format

## Progress
- [x] Allowed reading of new Xcode 11 xcresult file
- [x] Run `xcrun` commands to get the JSON output
- [x] Run on small sample project to get mapping of ☝️ to match Xcode 10 junit file
- [x] Clean up logic from ☝️ 
- [x] Link up failure files and line number and error messages
- [ ] Test on bigger test suites
- [x] Write tests

### Files changed
- Edited [test_parser.rb](https://github.com/xcpretty/trainer/pull/34/files#diff-7095933f509e386cc9028fc714dd6a3c)
- Added [xcresult.rb](https://github.com/xcpretty/trainer/pull/34/files#diff-8914ca65344f3666a46b17893c51eada)
- The majority of the 79 files changed here by adding an `.xcresult` file for tests

### Summary
- Added searching of path for `*.xcresult`
- Created all the models we needed based off of `xcrun xcresulttool formatDescription` in `xcresult.rb` 
- Gets JSON formatted results with  `xcrun xcresulttool get --format json --path #{path}`
  - Used to get an id of all the test summaries
  - Also used to get list of all the failures
- Uses ids of test summaries to get specific info for each test summary with `xcrun xcresulttool get --format json --path #{path} --id #{id}`
  - Used to get list, status, and duration of all tests run
  - Maps failures based upon target name and test name

### From Xcode 10
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="7" failures="2">
    <testsuite name="TestUITests" tests="1" failures="0" time="17.92351496219635">
        <testcase classname="TestUITests" name="testExample()" time="17.92015504837036">
        </testcase>
    </testsuite>
    <testsuite name="TestThisDude" tests="6" failures="2" time="0.7454169988632202">
        <testcase classname="TestTests" name="testExample()" time="0.0010919570922851562">
        </testcase>
        <testcase classname="TestTests" name="testFailureJosh1()" time="0.006746053695678711">
            <failure message="XCTAssertTrue failed (/Users/josh/Projects/fastlane/test-ios/TestTests/TestTests.swift:37)">
            </failure>
        </testcase>
        <testcase classname="TestTests" name="testPerformanceExample()" time="0.4788789749145508">
        </testcase>
        <testcase classname="TestThisDude" name="testExample()" time="0.00041794776916503906">
        </testcase>
        <testcase classname="TestThisDude" name="testFailureJosh2()" time="0.0015239715576171875">
            <failure message="XCTAssertTrue failed (/Users/josh/Projects/fastlane/test-ios/TestThisDude/TestThisDude.swift:36)">
            </failure>
        </testcase>
        <testcase classname="TestThisDude" name="testPerformanceExample()" time="0.2513129711151123">
        </testcase>
    </testsuite>
</testsuites>
```

### From Xcode 11
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="7" failures="2">
    <testsuite name="TestUITests" tests="1" failures="0" time="16.05245804786682">
        <testcase classname="TestUITests" name="testExample()" time="16.05245804786682">
        </testcase>
    </testsuite>
    <testsuite name="TestThisDude" tests="6" failures="2" time="0.5279300212860107">
        <testcase classname="TestTests" name="testExample()" time="0.0005381107330322266">
        </testcase>
        <testcase classname="TestTests" name="testFailureJosh1()" time="0.006072044372558594">
            <failure message="XCTAssertTrue failed (/Users/josh/Projects/fastlane/test-ios/TestTests/TestTests.swift#CharacterRangeLen=0&amp;EndingLineNumber=36&amp;StartingLineNumber=36)">
            </failure>
        </testcase>
        <testcase classname="TestTests" name="testPerformanceExample()" time="0.2661939859390259">
        </testcase>
        <testcase classname="TestThisDude" name="testExample()" time="0.0004099607467651367">
        </testcase>
        <testcase classname="TestThisDude" name="testFailureJosh2()" time="0.001544952392578125">
            <failure message="XCTAssertTrue failed (/Users/josh/Projects/fastlane/test-ios/TestThisDude/TestThisDude.swift#CharacterRangeLen=0&amp;EndingLineNumber=35&amp;StartingLineNumber=35)">
            </failure>
        </testcase>
        <testcase classname="TestThisDude" name="testPerformanceExample()" time="0.2531709671020508">
        </testcase>
    </testsuite>
</testsuites>
```

### CODE REVIEW

File: lib/trainer/test_parser.rb [hunk 1]
```diff
@@ -120,6 +132,97 @@ def test_group_and_name(testable_summary, test, xcpretty_naming)
       return group, name
     end
 
+    def parse_xcresult(path)
+      # Executes xcresulttool to get JSON format of the result bundle object
+      result_bundle_object_raw = `xcrun xcresulttool get --format json --path #{path}`
+      result_bundle_object = JSON.parse(result_bundle_object_raw)
```
Comment (snatchev) [inline lib/trainer/test_parser.rb]: if `xcrun` fails, I think it returns back "" (or maybe stderr?) which will cause a pretty cryptic JSON.parse exception.
Let's test `$?.success?` and raise something more friendly.

File: lib/trainer/test_parser.rb [hunk 2]
```diff
@@ -120,6 +132,97 @@ def test_group_and_name(testable_summary, test, xcpretty_naming)
...
+      rows = testable_summaries.map do |testable_summary|
+        all_tests = testable_summary.all_tests.flatten
+
+        row = {
+          project_path: testable_summary.project_relative_path,
+          target_name: testable_summary.target_name,
+          test_name: testable_summary.name,
+
+          duration: all_tests.map(&:duration).inject(:+),
+          tests: all_tests.map do |test|
```
Comment (snatchev) [inline lib/trainer/test_parser.rb]: let's set `tests` outside of the inline hash:

```
row[:tests] = all_tests.map ...
```

File: lib/trainer/test_parser.rb [hunk 3]
```diff
@@ -120,6 +132,97 @@ def test_group_and_name(testable_summary, test, xcpretty_naming)
...
+            # Tries to match failure on test case name
+            # Example:
+            # producing_target: "TestThisDude"
+            # test_case_name: "TestThisDude.testFailureJosh2()
+            found_failure = failures.find do |failure|
+              failure.test_case_name == test.identifier
+            end
+
+            # Set failure message if test status is "Failure" and a failure reference was found
+            if test.test_status == "Failure" && found_failure
```
Comment (snatchev) [inline lib/trainer/test_parser.rb]: Is there a good place to put some of this logic in the xcresult models? If it's a bad fit, then here is fine. It just seems like theres a lot of data that the test_parser needs to know about.

File: lib/trainer/test_parser.rb [hunk 4]
```diff
@@ -22,6 +22,8 @@ def self.auto_convert(config)
       files += Dir["#{containing_dir}/Test/*.xcresult/TestSummaries.plist"]
       files += Dir["#{containing_dir}/*.xcresult/TestSummaries.plist"]
       files += Dir[containing_dir] if containing_dir.end_with?(".plist") # if it's the exact path to a plist file
+      # Xcode 11
+      files += Dir["#{containing_dir}/*.xcresult"]
```
Comment (juliangrosshauser) [inline lib/trainer/test_parser.rb]: Do you mind adding the following globs, like for the previous Xcode versions?

```ruby
files += Dir["#{containing_dir}/**/Logs/Test/*.xcresult"]
files += Dir["#{containing_dir}/Test/*.xcresult"]
```

Comment (joshdholtz) [reply lib/trainer/test_parser.rb]: You got it!

**Comment by joshdholtz** [comment]
@forceunwrap Thanks for mentioning that! I fixed this PR for both of those 😊

**Comment by forceunwrap** [comment]
That was fast, thanks 😁

**Pull request merged**